blockchain/chain_sync/src/tipset_syncer.rs

```diff
@@ -0,0 +1,1476 @@
+use cid::Cid;
+use clock::ChainEpoch;
+use crypto::{verify_bls_aggregate, DomainSeparationTag};
+use encoding::Cbor;
+use fil_types::{
+    verifier::ProofVerifier, NetworkVersion, Randomness, ALLOWABLE_CLOCK_DRIFT, BLOCK_GAS_LIMIT,
+    TICKET_RANDOMNESS_LOOKBACK,
+};
+use forest_libp2p::chain_exchange::TipsetBundle;
+use interpreter::price_list_by_epoch;
+use ipld_blockstore::BlockStore;
+use message::{Message, UnsignedMessage};
+use networks::{get_network_version_default, BLOCK_DELAY_SECS, UPGRADE_SMOKE_HEIGHT};
+use state_manager::StateManager;
+use state_tree::StateTree;
+
+const MAX_TIPSETS_TO_REQUEST: u64 = 100;
+
+struct TipsetGroup {
+    tipsets: Vec<Arc<Tipset>>,
+    epoch: ChainEpoch,
+    parents: TipsetKeys,
+}
+
+impl TipsetGroup {
+    fn new(tipset: Arc<Tipset>) -> Self {
+        let epoch = tipset.epoch();
+        let parents = tipset.parents().clone();
+        Self {
+            tipsets: vec![tipset],
+            epoch,
+            parents,
+        }
+    }
+
+    fn epoch(&self) -> ChainEpoch {
+        self.epoch
+    }
+
+    fn parents(&self) -> TipsetKeys {
+        self.parents.clone()
+    }
+    // Attempts to add a tipset to the group
+    // If the tipset is added, the method returns `None`
+    // If the tipset is discarded, the method return `Some(tipset)`
+    fn try_add_tipset(&mut self, tipset: Arc<Tipset>) -> Option<Arc<Tipset>> {
+        // The new tipset must:
+        //  1. Be unique
+        //  2. Have the same epoch and parents as the other tipsets in the group
+        if !self.epoch.eq(&tipset.epoch()) || !self.parents.eq(tipset.parents()) {
+            return Some(tipset);
+        }
+        if let Some(_) = self.tipsets.iter().find(|ts| tipset.key().eq(&ts.key())) {
+            return Some(tipset);
+        }
+        self.tipsets.push(tipset);
+        None
+    }
+
+    fn take_heaviest_tipset(&mut self) -> Option<Arc<Tipset>> {
+        self.tipsets
+            .iter()
+            .enumerate()
+            .max_by_key(|(idx, ts)| ts.weight())
+            .map(|(idx, _)| idx)
+            .map(|idx| self.tipsets.swap_remove(idx))
+    }
+
+    fn heaviest_weight(&self) -> BigInt {
+        // Unwrapping is safe because we initialize the struct with at least one tipset
+        self.tipsets
+            .iter()
+            .map(|ts| ts.weight().clone())
+            .max()
+            .unwrap()
+    }
+
+    fn merge(&mut self, other: Self) {
+        other.tipsets.into_iter().for_each(|ts| {
+            self.try_add_tipset(ts);
+        });
+    }
+
+    fn is_mergeable(&self, other: &Self) -> bool {
+        self.epoch.eq(&other.epoch) && self.parents.eq(&other.parents)
+    }
+
+    fn is_heavier_than(&self, other: &Self) -> bool {
+        self.heaviest_weight() > other.heaviest_weight()
+    }
+
+    fn tipsets(self) -> Vec<Arc<Tipset>> {
+        self.tipsets
+    }
+}
+
+/// The StreamingTipsetSyncer receives and triages a stream of Tipsets
```

Comment:
Might wanna describe what "triages" means exactly or use another word for this. Also `StreamingTipsetSyncer`?